### PR TITLE
Expose the status of currently-running tasks

### PR DESCRIPTION
**Pull request by cnr**
Fixes https://github.com/fossas/team-analysis/issues/418 . Details available in that ticket

The user-visible changes for this include:

- The `Diagnostics` `context` stack is visible at the bottom of the terminal for each running task (when the terminal is ANSI-compatible)
- `--debug` mode (very) noisily outputs each `context` change for each task as discrete log messages (this works even when the terminal is not ANSI-compatible)
- Each running task displays a unique integer ID
- Each analyzer has `context` added for:
  - Project discovery
  - Project analysis
    - the type of analysis; dynamic/static; and specific analysis strategy names as appropriate
    - for complex analyzers (e.g. the golang analyzers that attempt to use a command for deep dependency analysis), it should be clear from the `context` what step is currently running in the analyzer
- `Exec`
  - has `context` for running commands
- `ReadFS`
  - has `context` for parsing various types of files
  - has `context` for walking the filetree

---

This PR also addresses some tech debt:
- Sticky regions in the console should work better now on all platforms, but especially windows: we're using the `concurrent-output` package for this
- Migrates other entrypoints to use `logWithExit` helper
- Adds convenient string conversion functions in `Data.String.Conversion`
  - This eliminates our use of the unsafe `decodeUtf8` functions and replaces them with `decodeUtf8With lenientDecode`
- Eliminate `ResultBundle` from `Diagnostics`, as the warnings from the bundle were never actually used (and it just made for a worse dev experience)

### CODE REVIEW

File: src/Strategy/Gomodules.hs
Comment (scruffystuffs) [inline src/Strategy/Gomodules.hs:49]: I'm not sure everyone who works on non-scaffolding areas of the code will immediately understand how these multiple, nested `context`s are going to compose.  Some questions they might have immediately:
* Are contexts additive?
* When and how are they dropped and replaced?
* Are they cleared if not replaced by something else?
 
I think the semantics need to be clear to us without trying to make everyone else figure out how haskell is going to handle the effects.  While you need haskell expertise to hack through the scaffolding of this project, I think we should try very hard to make that less true for the business logic, and strategies in particular should be as close to easy as possible.

The solution is dev docs, in my opinion.  I would really like to see how it resolves in frames, i.e. showing a reader the expected printed value, explaining when it moves to the next value, and how the value will be expected to change.

**Review by scruffystuffs — COMMENTED** [review]
Overall, I think the approach is great and I'm happy with almost all of the changes.
I am still working on the full in-depth review, but I had some initial feedback about the API for non-diagnostic stickies, and a comment about docs for diagnostic contexts

File: src/Strategy/Gomodules.hs
Comment (cnr) [reply src/Strategy/Gomodules.hs:49]: Good question -- the semantics of `context` for Diagnostics is that it's additive and scoped:
```hs
context "foo" $ do
  context "bar" $ do
    -- context is `foo -> bar` here
    pure ()
  context "baz" $ do
    -- context is `foo -> baz` here
    pure ()
  -- context is `foo` here
```

here's an example of a deeply nested context stack during gradle discovery:
```
[TASK 6] Gradle > Finding projects > Walking the filetree > Listing gradle projects at '/Users/connor/Desktop/tmp2/grpc-java/' > Running command '/Users/connor/Desktop/tmp2/grpc
-java/gradlew'
```

I'll add more docs to `context` to make that clear. The primary purpose for `context` is to create "traceback"s for thrown errors:

```
[ WARN] ----------
  An error occurred:

      Command execution failed:
          command: gradle
          args: [projects]
          dir: /Users/connor/Desktop/tmp2/grpc-java/
          exit: ExitFailure 1
          stdout:

          stderr:
            gradle: startProcess: runInteractiveProcess: exec: does not exist (No such file or directory)

      Traceback:
        - Running command 'gradle'
        - Listing gradle projects at '/Users/connor/Desktop/tmp2/grpc-java/'
```

..and we're just hijacking it to produce status messages

File: src/App/Fossa/API/BuildWait.hs
Comment (cnr) [reply src/App/Fossa/API/BuildWait.hs:50]: Good catch. I'd initially experienced issues with duplicate `StickyLogger` effects when there were multiple sticky regions -- but I agree the approach of using an effect is the better one (and `fused-effects` has machinery for overlapping effects through `Control.Effect.Labelled`)

File: src/Control/Carrier/Diagnostics/StickyContext.hs
Comment (scruffystuffs) [inline src/Control/Carrier/Diagnostics/StickyContext.hs:46]: [question]: Am I reading this right?

I think what's happening is that the `alg`ebra for this effect is just special-casing `Context` effects.  Which means that `L somethingElse -> ...` is a pass-through to normal diag stuff, and `R other -> ...` is the same, but also passing through the three readers of this effect via `(R (R (R other)))`.

**Review by scruffystuffs — COMMENTED** [review]
A few more nits, but I'm satisfied that the only major concern is the sticky API comment I left first.

Once some of my questions are answered, I can be more confident that I do understand each part, and I can probably test and approve quickly after that.  

Awesome work!  Super creative, idiomatic in a lot of ways, really good API decisions, and solid code.

File: src/App/Fossa/API/BuildWait.hs
Comment (scruffystuffs) [reply src/App/Fossa/API/BuildWait.hs:50]: I think I imagined that the `StickyLogger` as I proposed is for one-region use only.  In most of these cases, that seems to be true.  When it's not, I think we need to bring context around in the form of diagnostics + sticky, using the new `SitckyDiag` you already created.

File: src/Strategy/Gomodules.hs
Comment (zlav) [reply src/Strategy/Gomodules.hs:49]: I think the use of context here is slightly confusing, but it does make sense and the resulting errors are very nice looking so I'm happy about that. As long as we document this I am ok with it.

File: src/Control/Effect/Diagnostics.hs
Comment (cnr) [reply src/Control/Effect/Diagnostics.hs:54]: The prime is meant to match the definition of `recover'` below; we're keeping the previous `recover`, which is just a thin wrapper around `recover'`

The naming convention I've been using is to add a `'` to denote the "more-complicated" variants of functions

File: src/Data/String/Conversion.hs
Comment (cnr) [reply src/Data/String/Conversion.hs:112]: Yeah, this is one of the weird scenarios where if you know what the extension does, it's obvious why it's being used here -- and if you don't know what the extension does, you can probably just ignore it


Kind of conflicted as to whether it deserves specific docs, though I'll add some this time

File: src/Control/Effect/Diagnostics.hs
Comment (scruffystuffs) [reply src/Control/Effect/Diagnostics.hs:54]: TLDR: it's up to you, I'm totally fine either way.

> The naming convention I've been using is to add a ' to denote the "more-complicated" variants of functions

If that's the convention we're using, then I think we should usually remove prime marks unless more than one variant exists, or there's an exceptional reason.

I'm actually fine with either keeping or removing this particular prime, since it's not really necessary on its own, but it does match the effect function.  My preference is to remove it, since there's no real need to connect the names, and primes without variants add noise, but I'm probably not going to remember it's even there once this merges.

File: src/Data/String/Conversion.hs
Comment (scruffystuffs) [reply src/Data/String/Conversion.hs:112]: I think this one does deserve it: when I looked up `FunctionalDependencies`, I understood what was happening, but I didn't know **why** it was necessary.  That took a lot of thought about the extension and the class, which I think we can save others from doing later.

File: src/Control/Effect/Diagnostics.hs
Comment (cnr) [reply src/Control/Effect/Diagnostics.hs:94]: Agreed. I'll put together some docs. `errorBoundary` is primarily a hack, and is otherwise analogous to `runDiagnostics`.

`errorBoundary` is a workaround to allow us to send the `Diagnostics` constraint upward, so we can intercept `context` calls from the inner action. We use this for `stickyDiag` specifically

File: src/Effect/Logger.hs
Comment (cnr) [reply src/Effect/Logger.hs:82]: This is unfortunately required, because `displayConsoleRegions` doesn't guarantee flushed output when the action completes. I was seeing occurrences where stdout output (e.g. `--output` mode in `fossa analyze`) was getting dropped

The library has slightly frustrating UI, because `withConcurrentOutput` _does_ guarantee a flush, but you can't use `withConcurrentOutput` and `displayConsoleRegions` at the same time 😬

File: src/Strategy/Maven.hs
Comment (cnr) [reply src/Strategy/Maven.hs:53]: Hoping we switch to fourmolu 🙏

File: src/Control/Carrier/Fresh.hs
Comment (cnr) [reply src/Control/Carrier/Fresh.hs:4]: I used the same name as fused-effects: https://github.com/fused-effects/fused-effects/blob/master/src/Control/Effect/Fresh.hs

..but I think you're right that `AtomicCounter` is better

File: src/Control/Carrier/Diagnostics/StickyContext.hs
Comment (cnr) [reply src/Control/Carrier/Diagnostics/StickyContext.hs:46]: Yes, that's exactly right

File: src/Control/Effect/Diagnostics.hs
Comment (scruffystuffs) [inline src/Control/Effect/Diagnostics.hs:8]: [nit]: Can we keep the language pragmas together?

File: src/Effect/Logger.hs
Comment (scruffystuffs) [inline src/Effect/Logger.hs:3]: [nit]: Is this used here?  I couldn't find it with a text search.

**Review by scruffystuffs — APPROVED** [review]
One future concern and a few nits, but this LGTM!
giphub

**Pull request merged**
